src/app.js

```diff
@@ -221,30 +216,128 @@ export default function (app) {
+      var reusableChildren = {}
+      var oldElements = []
+      var newKeys = {}
+
+
+      for (var i = 0; i < oldLen; i++) {
+        var oldElement = element.childNodes[i]
+        oldElements[i] = oldElement
 
-      for (var i = 0; i < len || i < oldLen; i++) {
-        patch(element, element.childNodes[i], oldNode.children[i], node.children[i])
+        var oldChild = oldNode.children[i]
+        var oldKey = getKeyFrom(oldChild)
+
+        if (null != oldKey) {
+          reusableChildren[oldKey] = [oldElement, oldChild]
+        }
       }
 
+      var i = 0
+      var j = 0
+
+      while (j < len) {
+        var oldElement = oldElements[i]
+        var oldChild = oldNode.children[i]
+        var newChild = node.children[j]
+
+        var oldKey = getKeyFrom(oldChild)
+        if (newKeys[oldKey]) {
+          i++
+          continue
+        }
+
+        var newKey = getKeyFrom(newChild)
+
+        var reusableChild = reusableChildren[newKey]
+        var reusableElement = 0
+        var reusableNode = 0
+
+        if (reusableChild) {
+          reusableElement = reusableChild[0]
+          reusableNode = reusableChild[1]
+        }
+
+        if (null == oldKey && null == newKey) {
+          patch(element, oldElement, oldChild, newChild)
+          j++
+          i++
+
+        } else if (null == oldKey && null != newKey) {
+          if (reusableElement) {
+            element.insertBefore(reusableElement, oldElement)
+            patch(element, reusableElement, reusableNode, newChild)
+          } else {
+            patch(element, oldElement, null, newChild)
+          }
+
+          j++
+          newKeys[newKey] = newChild
+
+        } else if (null != oldKey && null == newKey) {
+          i++
+
+        } else {
+          if (oldKey === newKey) {
+            patch(element, reusableElement, reusableNode, newChild)
+            i++
+
+          } else if (reusableElement) {
+            element.insertBefore(reusableElement, oldElement)
+            patch(element, reusableElement, reusableNode, newChild)
+
+          } else {
+            patch(element, oldElement, null, newChild)
+          }
+
+          j++
+          newKeys[newKey] = newChild
+        }
+      }
+
+      while (i < oldLen) {
+        var oldChild = oldNode.children[i]
+        var oldKey = getKeyFrom(oldChild)
+        if (null == oldKey) {
+          batchRemove(element, oldElements[i], oldChild)
+        }
+        i++
+      }
+
+      for (var i in reusableChildren) {
+        var reusableChild = reusableChildren[i]
+        var reusableNode = reusableChild[1]
+        if (!newKeys[reusableNode.data.key]) {
+          batchRemove(element, reusableChild[0], reusableNode)
+        }
+      }
     } else if (node !== oldNode) {
       var i = element
       parent.replaceChild(element = createElementFrom(node), i)
     }
 
     return element
   }
+
+  function getKeyFrom(node) {
+    if (node && (node = node.data)) {
```

Comment:
Yes let's :) I'm not at all sure it will make a big difference, but I'm hearing this sort of recommendation a lot of late